### ISSUE TITLE
Setup release-monitor for fleetdm/fleet

### ISSUE THREAD

**Issue by varunsh-coder** [step-security/secure-workflows#671]
This issue is to track setting up `release-monitor` for [fleetdm/fleet](https://github.com/fleetdm/fleet). The expected release KB for fleet GitHub release is [here](https://github.com/step-security/secure-workflows/blob/main/release-knowledge-base/github/release/fleetdm/fleet/release.yml).
 
@zwass, following up from our discussion [here](https://github.com/step-security/harden-runner/issues/123#issuecomment-1118142359), we are planning to store `expected release workflows` as YAML files in this repo. 

[Here](https://github.com/step-security/secure-workflows/blob/main/release-knowledge-base/github/release/fleetdm/fleet/release.yml) is the one for fleet. There are a few others for other scenarios, e.g. npm, maven etc. Please share feedback on the YAML format. It is expected to evolve over time. https://github.com/step-security/secure-workflows/tree/main/release-knowledge-base

1. w.r.t patch releases, would you prefer to get notified every time release is not from `main` branch, or do you want to setup a process, e.g. branch must start with `release*`, and you can then setup branch protection rules similar to `main` for such patch branches. 

2. w.r.t notification, if release is not via expected process, do you want to be 
  a. Emailed or 
  b. Should we create an issue in the repo, or 
  c. Do you want to receive a web hook?

As an example, I will setup email and web hook for step-security/agent. With web hook I should be able to configure slack notification/ get a phone call via other integrations.

**Comment by zwass** [comment]
I'd prefer webhook or issue in the repo. If we can set a branch prefix allowed for patch releases then that does cover our use case.

**Comment by varunsh-coder** [comment]
Hi @zwass, the first release of the `release-monitor` is ready. Please give it a try and let me know your feedback. You can install the App from here: https://github.com/apps/stepsecurity-app. Once you install it on the `fleetdm/fleet` repository, every time there is a release, it will check if the [release process](https://github.com/step-security/secure-workflows/blob/main/release-knowledge-base/github/release/fleetdm/fleet/release.yml) was followed. 

If not, an [issue like this](https://github.com/harden-runner-canary/agent/issues/14) will be created. As of now, we only have `main` as the expected branch for `fleetdm/fleet`, so if there is a release from a different branch, you should get such an issue created. 

We are working on the re-build from source part, and will update you once that is done. That will re-build fleetdm/fleet and compare the checksum to what is in the `release asset`. 

Also, I noticed that if the `release asset` is modified, it does not trigger a webhook, so we need to probably poll periodically to check changes to release assets...

**Comment by zwass** [comment]
I've installed the app, so it sounds like we are set for now?

Does the branch name logic support wildcards? Does it support tags rather than branch names?

Thank you!

**Comment by varunsh-coder** [comment]
> I've installed the app, so it sounds like we are set for now?
> 
> Does the branch name logic support wildcards? Does it support tags rather than branch names?
> 
> Thank you!

Thanks @zwass! Yes, you are set for now for the following checks:
1. Release was from expected workflow
2. Release commit is in the expected branch

Branch name logic will support wildcards in a couple of days. I will reach out to setup wildcard for patch releases. 

It does not support tag names. How would you use tags if it was supported? Would it be to check if the tag name follows an expected pattern?

**Comment by zwass** [comment]
Yep, I'd be looking for the same pattern matching logic for tags as branch names.

**Comment by varunsh-coder** [comment]
Hi @zwass, the support for wildcards has been added to branch names. Please let me know what pattern you want to set for branch name for patch releases, and I can update the `expected release process` YAML file. Also, what pattern were you thinking of setting for tag names? Will start work on pattern matching logic for tags soon. 

We have received feedback that the `expected release process` YAML file should be allowed in the default branch of the repository being monitored. That will allow maintainers to update it in their own repo. Please let me know if you have any thoughts on this. We can have a hierarchy, where it checks in the repo being monitored first, and if not found, checks in step-security/secure-workflows...

**Comment by varunsh-coder** [comment]
Support for tags has been added. [This](https://github.com/step-security/secure-workflows/blob/main/knowledge-base/releases/github/release/fleetdm/fleet/release-monitor.yml) is the updated `expected release process` YAML file for Fleet.

**Comment by zwass** [comment]
I would love to see a status if GitHub had a notion of "Checks" for releases. Where do you think think the status could go? Maybe it could be a webhook and then we could hook it up to our Slack?

**Comment by varunsh-coder** [comment]
> I would love to see a status if GitHub had a notion of "Checks" for releases. Where do you think think the status could go? Maybe it could be a webhook and then we could hook it up to our Slack?

Webhook seems like a good idea. As next step, we are working on reproducing the build. Will keep you posted once that is done.

**Comment by varunsh-coder** [comment]
We are able to reproduce builds for `fleetctl` and match the checksums in the released artifacts! 
The [release-monitor.yml](https://github.com/step-security/secure-workflows/blob/main/knowledge-base/releases/github/release/fleetdm/fleet/release-monitor.yml) file has been updated to add instructions to reproduce builds. 

These the the reproduced builds for `fleetctl` for linux, mac, and windows. The [workflow](https://github.com/step-security/reproduce-builds/blob/main/.github/workflows/rebuild.yml) is setup to pass the build if checksum matches expected checksum. 
https://github.com/step-security/reproduce-builds/runs/6782518145?check_suite_focus=true
https://github.com/step-security/reproduce-builds/runs/6782514700?check_suite_focus=true
https://github.com/step-security/reproduce-builds/runs/6782513499?check_suite_focus=true

Unfortunately, we are not able to reproduce builds for `fleet`. The same build command results in different output each time it is run. 

Here are couple of runs that result in different checksums:
https://github.com/step-security/reproduce-builds/runs/6782512111?check_suite_focus=true
https://github.com/step-security/reproduce-builds/runs/6782387743?check_suite_focus=true

Let me know if you know why this may be the case, or how to make build reproducible for `fleet`. Thanks!

**Comment by zwass** [comment]
I assume the build for Fleet is non-reproducible because it uses CGO? Not sure what it will take to make that reproducible.

**Comment by varunsh-coder** [comment]
> I assume the build for Fleet is non-reproducible because it uses CGO? Not sure what it will take to make that reproducible.

I will investigate and check how other projects with CGO setup reproducible builds.

**Comment by varunsh-coder** [comment]
@h0x0er has made good progress to make `fleet` build reproducible. Looks like we need to add one or two more flags. @h0x0er please share your learnings in this thread once you complete your investigation. We can then discuss making those changes to Fleet's goreleaser file.
